pkg/controller/unreachable/unreachable_controller_test.go

```diff
@@ -0,0 +1,239 @@
+  name: bar
+`
+	testRemoteClusterCurrentVersion = "4.0.0"
+	remoteClusterVersionObjectName  = "version"
+
+	remoteClusterRouteObjectName      = "console"
+	remoteClusterRouteObjectNamespace = "openshift-console"
+)
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+}
+
+func TestUnreachableClusterDeployment(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+	tests := []struct {
+		name        string
+		existing    []runtime.Object
+		expectError bool
+		validate    func(*testing.T, *hivev1.ClusterDeployment)
+	}{
+		{
+			name: "unreachable condition should be added",
+			existing: []runtime.Object{
+				testClusterDeployment(),
+				testKubeconfigSecret(),
+			},
+			validate: func(t *testing.T, cd *hivev1.ClusterDeployment) {
+				cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition)
+				if cond.Status != corev1.ConditionTrue {
+					t.Errorf("Did not get expected state for unreachable condition. Expected: \n%#v\nGot: \n%#v", corev1.ConditionTrue, cond.Status)
+				}
+			},
+		},
+		{
+			name: "unreachable condition should change from false to true",
+			existing: []runtime.Object{
+				addFalseUnreachableCondition(),
+				testKubeconfigSecret(),
+			},
+			validate: func(t *testing.T, cd *hivev1.ClusterDeployment) {
+				cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition)
+				if cond.Status != corev1.ConditionTrue {
+					t.Errorf("Did not get expected state for unreachable condition. Expected: \n%#v\nGot: \n%#v", corev1.ConditionTrue, cond.Status)
+				}
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			fakeClient := fake.NewFakeClient(test.existing...)
+			rcd := &ReconcileRemoteMachineSet{
+				Client:                        fakeClient,
+				scheme:                        scheme.Scheme,
+				logger:                        log.WithField("controller", "unreachable"),
+				remoteClusterAPIClientBuilder: testRemoteClusterAPIClientBuilder,
+			}
+
+			namespacedName := types.NamespacedName{
+				Name:      testName,
+				Namespace: testNamespace,
+			}
+
+			_, err := rcd.Reconcile(reconcile.Request{NamespacedName: namespacedName})
+
+			if test.validate != nil {
+				cd := &hivev1.ClusterDeployment{}
+				err := fakeClient.Get(context.TODO(), namespacedName, cd)
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+					return
+				}
+				test.validate(t, cd)
+			}
+
+			if err != nil && !test.expectError {
+				t.Errorf("Unexpected error: %v", err)
+			}
+			if err == nil && test.expectError {
+				t.Errorf("Expected error but got none")
+			}
+		})
+	}
+}
+
+func TestIsTimeForConnectivityCheck(t *testing.T) {
+	cd := addFalseUnreachableCondition()
+	cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition)
+	if !isTimeForConnectivityCheck(cond) {
+		t.Error("Exppected isTimeForConnectivityCheck() to return true")
+	}
```

Comment:
Lets hit this function a little harder, it's a bit tricky and I'd like to see it returning true or false in a wide variety of situations.

Reply:
Agree